File: backend-nestjs/src/common/events/commission-events.ts
/**
 * Event Contract: Commission Module ↔ TPA Core
 * Version: 1.0.0
 *
 * This file defines the event schemas for communication between
 * the Medical Commission module (NestJS) and TPA Core engine (Go).
 *
 * Events are:
 * - Versioned (schema version in each event)
 * - Immutable (once published, never modified)
 * - Auditable (event_id, timestamp, correlation)
 * - Transport-agnostic (JSON)
 */

export const EVENT_VERSION = '1.0.0';

// Event Types
export enum EventType {
  // Commission case events
  COMMISSION_CASE_CREATED = 'commission.case.created',
  COMMISSION_CASE_ASSIGNED = 'commission.case.assigned',
  COMMISSION_CASE_REVIEWED = 'commission.case.reviewed',
  COMMISSION_VERDICT_ISSUED = 'commission.verdict.issued',
  COMMISSION_CASE_CLOSED = 'commission.case.closed',
  COMMISSION_CASE_APPEALED = 'commission.case.appealed',

  // Social work events
  SOCIALWORK_CASE_CREATED = 'socialwork.case.created',
  SOCIALWORK_ASSESSMENT_DONE = 'socialwork.assessment.done',
  SOCIALWORK_REFERRAL_ISSUED = 'socialwork.referral.issued',
}

// Base event interface
export interface BaseEvent {
  event_id: string; // UUID
  event_type: EventType;
  version: string;
  timestamp: string; // ISO 8601
  source: string; // 'commission-api' | 'tpa-api'

  // Correlation
  correlation_id?: string;
  causation_id?: string;

  // Tenant context
  tenant_id: number;
}

// Insured person reference
export interface InsuredPersonRef {
  id: string;
  national_id: string;
  personnel_code: string;
  full_name: string;
  relation: 'main' | 'spouse' | 'child' | 'parent';
}

// Verdict details
export interface VerdictDetails {
  verdict_id: string;
  verdict_type: string;
  verdict_code: string;
  verdict_text: string;
  disability_rate?: number; // 0-100
  effective_from?: string;
  effective_to?: string;
  is_permanent: boolean;
  needs_review: boolean;
  review_date?: string;
  approved_by: string;
  approved_at: string;
  commission_level: 'provincial' | 'central';
}

// Financial impact
export interface FinancialImpact {
  coverage_type: 'full' | 'partial' | 'none';
  coverage_percent: number;
  coverage_limit_delta: number;

  monthly_allowance?: number;
  lump_sum_payment?: number;

  service_restrictions?: string[];
  provider_restrictions?: string[];

  valid_from: string;
  valid_to?: string;
}

// Document reference
export interface DocumentRef {
  document_id: string;
  document_type: 'verdict_letter' | 'medical_report' | 'attachment';
  file_name: string;
  file_url: string;
  checksum: string;
}

// Commission Verdict Event
export interface CommissionVerdictEvent extends BaseEvent {
  event_type: EventType.COMMISSION_VERDICT_ISSUED;

  case_id: string;
  case_number: string;

  insured_person: InsuredPersonRef;
  verdict: VerdictDetails;
  financial_impact?: FinancialImpact;
  documents?: DocumentRef[];
}

// Referral details for social work
export interface ReferralDetails {
  referral_id: string;
  referral_type: string;
  referral_reason: string;
  referred_to: string;
  priority: 'normal' | 'urgent' | 'critical';
  due_date?: string;
  amount?: number;
  issued_by: string;
  issued_at: string;
  notes?: string;
}

// Social Work Referral Event
export interface SocialWorkReferralEvent extends BaseEvent {
  event_type: EventType.SOCIALWORK_REFERRAL_ISSUED;

  case_id: string;
  case_type: string;

  insured_person: InsuredPersonRef;
  referral: ReferralDetails;
}

// Event envelope for transport
export interface EventEnvelope<T extends BaseEvent> {
  event: T;
  schema: string;
  encoded: boolean;
}

// Helper to create events
export function createEvent<T extends BaseEvent>(
  type: EventType,
  tenantId: number,
  payload: Omit<T, keyof BaseEvent>,
): T {
  const base: BaseEvent = {
    event_id: crypto.randomUUID(),
    event_type: type,
    version: EVENT_VERSION,
    timestamp: new Date().toISOString(),
    source: 'commission-api',
    tenant_id: tenantId,
  };

  return { ...base, ...payload } as T;
}

// Validation
export function validateEvent(event: BaseEvent): string[] {
  const errors: string[] = [];

  if (!event.event_id) errors.push('event_id is required');
  if (!event.tenant_id) errors.push('tenant_id is required');
  if (!event.event_type) errors.push('event_type is required');
  if (!event.timestamp) errors.push('timestamp is required');

  return errors;
}

File: backend-go/internal/domain/event/commission_events.go
package event

import (
	"encoding/json"
	"time"
)

// EventVersion defines the current event schema version
const EventVersion = "1.0.0"

// EventType defines event types for commission decisions
type EventType string

const (
	// Commission case events
	EventCommissionCaseCreated    EventType = "commission.case.created"
	EventCommissionCaseAssigned   EventType = "commission.case.assigned"
	EventCommissionCaseReviewed   EventType = "commission.case.reviewed"
	EventCommissionVerdictIssued  EventType = "commission.verdict.issued"
	EventCommissionCaseClosed     EventType = "commission.case.closed"
	EventCommissionCaseAppealed   EventType = "commission.case.appealed"

	// Social work events
	EventSocialWorkCaseCreated    EventType = "socialwork.case.created"
	EventSocialWorkAssessmentDone EventType = "socialwork.assessment.done"
	EventSocialWorkReferralIssued EventType = "socialwork.referral.issued"
)

// BaseEvent contains common fields for all events
type BaseEvent struct {
	// Event metadata
	EventID     string    `json:"event_id"`      // Unique event identifier (UUID)
	EventType   EventType `json:"event_type"`    // Type of event
	Version     string    `json:"version"`       // Schema version
	Timestamp   time.Time `json:"timestamp"`     // When event occurred
	Source      string    `json:"source"`        // Source service (commission-api, tpa-api)

	// Correlation
	CorrelationID string `json:"correlation_id,omitempty"` // For tracing related events
	CausationID   string `json:"causation_id,omitempty"`   // Event that caused this event

	// Tenant context
	TenantID uint `json:"tenant_id"`
}

// CommissionVerdictEvent is emitted when a medical commission issues a verdict
// This is the primary event consumed by TPA Core for financial decisions
type CommissionVerdictEvent struct {
	BaseEvent

	// Case reference
	CaseID     string `json:"case_id"`      // UUID of commission case
	CaseNumber string `json:"case_number"`  // Human-readable case number

	// Insured person
	InsuredPerson InsuredPersonRef `json:"insured_person"`

	// Verdict details
	Verdict VerdictDetails `json:"verdict"`

	// Financial implications (to be processed by TPA Core)
	FinancialImpact *FinancialImpact `json:"financial_impact,omitempty"`

	// Attachments
	Documents []DocumentRef `json:"documents,omitempty"`
}

// InsuredPersonRef contains reference to insured person
type InsuredPersonRef struct {
	ID            string `json:"id"`             // UUID in commission system
	NationalID    string `json:"national_id"`    // کد ملی
	PersonnelCode string `json:"personnel_code"` // کد پرسنلی
	FullName      string `json:"full_name"`
	Relation      string `json:"relation"`       // main, spouse, child, parent
}

// VerdictDetails contains the commission verdict
type VerdictDetails struct {
	VerdictID       string     `json:"verdict_id"`
	VerdictType     string     `json:"verdict_type"`     // disability, disease, etc.
	VerdictCode     string     `json:"verdict_code"`     // Standard verdict code
	VerdictText     string     `json:"verdict_text"`     // Full verdict text
	DisabilityRate  *int       `json:"disability_rate"`  // درصد از کار افتادگی (0-100)
	EffectiveFrom   *time.Time `json:"effective_from"`
	EffectiveTo     *time.Time `json:"effective_to"`
	IsPermanent     bool       `json:"is_permanent"`
	NeedsReview     bool       `json:"needs_review"`
	ReviewDate      *time.Time `json:"review_date,omitempty"`
	ApprovedBy      string     `json:"approved_by"`      // User who approved
	ApprovedAt      time.Time  `json:"approved_at"`
	CommissionLevel string     `json:"commission_level"` // provincial, central
}

// FinancialImpact describes how this verdict affects coverage
type FinancialImpact struct {
	// Coverage changes
	CoverageType       string  `json:"coverage_type"`        // full, partial, none
	CoveragePercent    int     `json:"coverage_percent"`     // New coverage percentage
	CoverageLimitDelta int64   `json:"coverage_limit_delta"` // Change in coverage limit

	// Benefits
	MonthlyAllowance   int64   `json:"monthly_allowance,omitempty"`
	LumpSumPayment     int64   `json:"lump_sum_payment,omitempty"`

	// Restrictions
	ServiceRestrictions []string `json:"service_restrictions,omitempty"`
	ProviderRestrictions []string `json:"provider_restrictions,omitempty"`

	// Validity
	ValidFrom time.Time  `json:"valid_from"`
	ValidTo   *time.Time `json:"valid_to,omitempty"`
}

// DocumentRef references an attached document
type DocumentRef struct {
	DocumentID   string `json:"document_id"`
	DocumentType string `json:"document_type"` // verdict_letter, medical_report, etc.
	FileName     string `json:"file_name"`
	FileURL      string `json:"file_url"`
	Checksum     string `json:"checksum"` // For integrity verification
}

// SocialWorkReferralEvent is emitted when social work issues a referral
type SocialWorkReferralEvent struct {
	BaseEvent

	// Case reference
	CaseID   string `json:"case_id"`
	CaseType string `json:"case_type"` // financial_aid, loan, equipment, etc.

	// Insured person
	InsuredPerson InsuredPersonRef `json:"insured_person"`

	// Referral details
	Referral ReferralDetails `json:"referral"`
}

// ReferralDetails contains referral information
type ReferralDetails struct {
	ReferralID     string    `json:"referral_id"`
	ReferralType   string    `json:"referral_type"`
	ReferralReason string    `json:"referral_reason"`
	ReferredTo     string    `json:"referred_to"`      // Department/unit
	Priority       string    `json:"priority"`         // normal, urgent, critical
	DueDate        *time.Time `json:"due_date,omitempty"`
	Amount         *int64    `json:"amount,omitempty"` // If financial
	IssuedBy       string    `json:"issued_by"`
	IssuedAt       time.Time `json:"issued_at"`
	Notes          string    `json:"notes,omitempty"`
}

// EventEnvelope wraps events for transport
type EventEnvelope struct {
	Event   interface{} `json:"event"`
	Schema  string      `json:"schema"`  // JSON Schema reference
	Encoded bool        `json:"encoded"` // If event is base64 encoded
}

// ToJSON serializes event to JSON
func (e *CommissionVerdictEvent) ToJSON() ([]byte, error) {
	return json.Marshal(e)
}

// FromJSON deserializes event from JSON
func (e *CommissionVerdictEvent) FromJSON(data []byte) error {
	return json.Unmarshal(data, e)
}

// Validate validates event data
func (e *CommissionVerdictEvent) Validate() error {
	if e.EventID == "" {
		return ErrMissingEventID
	}
	if e.TenantID == 0 {
		return ErrMissingTenantID
	}
	if e.CaseID == "" {
		return ErrMissingCaseID
	}
	if e.InsuredPerson.NationalID == "" {
		return ErrMissingNationalID
	}
	return nil
}

// Event validation errors
var (
	ErrMissingEventID    = eventError("event_id is required")
	ErrMissingTenantID   = eventError("tenant_id is required")
	ErrMissingCaseID     = eventError("case_id is required")
	ErrMissingNationalID = eventError("national_id is required")
)

type eventError string

func (e eventError) Error() string { return string(e) }
